### PR TITLE
[ENUM] Declare DLDeviceType as int32  enum in c++

### PR DESCRIPTION
**Pull request by tqchen**

### CODE REVIEW

**Comment by tqchen** [comment]
Helps to further clarify the specification

File: include/dlpack/dlpack.h [hunk 1]
```diff
@@ -26,6 +26,12 @@
 #define DLPACK_DLL
 #endif
 
+#ifdef __cplusplus
+#define DLPACK_INT32_ENUM_TYPE : int32_t
```
Comment (oleksandr-pavlyk) [inline include/dlpack/dlpack.h:30]: Perhaps `#include <cstdint>` before using `int32_t`.

**Review by oleksandr-pavlyk — APPROVED** [review]
Took me a minute to parse the change. Compiled with C++ compiler to be sure too.

File: include/dlpack/dlpack.h [hunk 2]
```diff
@@ -172,7 +178,7 @@ typedef struct {
   /*! \brief The device of the tensor */
   DLDevice device;
   /*! \brief Number of dimensions */
-  int ndim;
+  int32_t ndim;
```
Comment (csullivan) [inline include/dlpack/dlpack.h:181]: Could consider using `DLPACK_INT32_ENUM_TYPE` here to avoid future divergence.

Comment (csullivan) [reply include/dlpack/dlpack.h:181]: I believe we want to update [Lines 90/96](https://github.com/dmlc/dlpack/pull/86/files#diff-1564e5abd0da5cf46e82b28b0058996df944007494520a2b0a6b629880eaa097R96) as well.

`int device_id` -> `int32_t device_id` or equivalently 
-> `DLPACK_INT32_ENUM_TYPE device_id` as mentioned above.

**Review by csullivan — CHANGES_REQUESTED** [review]

File: include/dlpack/dlpack.h (hunk 2 above)
Comment (tqchen) [reply include/dlpack/dlpack.h:181]: in this case this is not an enum, mainly to clarify. `DLPACK_INT32_ENUM_TYPE` is indeed confusing, going to change `DLPACK_INT32_ENUM_ANNOTATION`

Comment (csullivan) [reply include/dlpack/dlpack.h:181]: In that case, I think the only change we would want is `int device_id -> int32_t device_id`.

**Review by csullivan — APPROVED** [review]

**Pull request merged**